### PR TITLE
Update to vergen 5, add branch, commit time, and build target to the panic metadata, automatically update app version from crate version

### LINKED ISSUE #1741: Upgrade to vergen 5
**Is your feature request related to a problem? Please describe.**

vergen 4 and 5 are breaking changes with a bunch of fixes, features, and removals.

**Describe the solution you'd like**

- [x] update to the new API
  - vergen rather than gen
  - Config rather than flags
- [x] update to the new env var names
- [x] replace our custom build script with a simple call to vergen (vergen now supports workspaces using libgit2, so we don't need that code any more)
- [ ] update the other vergen tickets with the new env var names

**Describe alternatives you've considered**

Do nothing, we have to maintain our own rerun-if-changed code.

### PR DESCRIPTION
**Pull request by fanatid**
## Motivation

Add more metadata with [vergen](https://crates.io/crates/vergen) crate.

## Solution

Update `vergen` to version 5.

## Review

I wanted add `rustc` metadata but this requires `sentry` update for the new version of `rustc_version`.

## Related Issues

- Require https://github.com/ZcashFoundation/zebra/pull/2026
- Closes https://github.com/ZcashFoundation/zebra/issues/1741
- Implements some of https://github.com/ZcashFoundation/zebra/issues/692

## Follow Up Work

Update `sentry` crate (require tokio 1 😢 + sentry-tracing update: https://github.com/kellpossible/sentry-tracing/pull/1)  and add `rustc` metadata.

### CODE REVIEW

File: zebrad/src/components/tracing/component.rs [hunk 1]
```diff
@@ -77,7 +77,7 @@ impl<A: abscissa_core::Application> Component<A> for Tracing {
     }
 
     fn version(&self) -> abscissa_core::Version {
-        abscissa_core::Version::parse("1.0.0-alpha.6").unwrap()
+        abscissa_core::Version::parse(env!("CARGO_PKG_VERSION")).unwrap()
```
Comment (teor2345) [inline zebrad/src/components/tracing/component.rs:80]: This change makes our release process a bit simpler.

So we can delete the "zebrad app code" from our release checklist:
https://github.com/ZcashFoundation/zebra/blob/main/.github/PULL_REQUEST_TEMPLATE/release-checklist.md#version-locations

File: zebrad/src/application.rs [hunk 2]
```diff
@@ -155,17 +155,19 @@ impl Application for ZebradApp {
...
         let panic_metadata = vec![
-            ("version", env!("CARGO_PKG_VERSION").to_string()),
-            ("git commit", Self::git_commit().to_string()),
-            ("Zcash network", network),
+            ("version", env!("CARGO_PKG_VERSION")),
+            ("git commit", Self::git_commit()),
+            ("commit timestamp", env!("VERGEN_GIT_COMMIT_TIMESTAMP")),
+            ("target triple", env!("VERGEN_CARGO_TARGET_TRIPLE")),
+            ("branch", env!("VERGEN_GIT_BRANCH")),
+            ("Zcash network", (&config.network.network).into()),
```
Comment (teor2345) [inline zebrad/src/application.rs:167]: Thanks I've been waiting for these for a while.

**Comment by teor2345** [comment]
> * Implements some of #692

I edited this line, because GitHub reads "partially resolve #ticket" as "resolve #ticket", and auto-closes the ticket.

**Comment by teor2345** [comment]
> * Require #2026

Yeah we really do need to move to the latest cc crate version, it's blocking a bunch of changes.

File: zebrad/src/components/tracing/component.rs (hunk 1 above)
Comment (teor2345) [reply zebrad/src/components/tracing/component.rs:80]: I added this change to the PR branch.

File: zebrad/src/application.rs (hunk 2 above)
Comment (teor2345) [inline zebrad/src/application.rs]: I just realised that the order of these variables is preserved in the panic `metadata_section`. So let's keep them roughly sorted, to make bug reports a bit easier to read:
```suggestion
            // git
            ("version", env!("CARGO_PKG_VERSION")),
            ("branch", env!("VERGEN_GIT_BRANCH")),
            ("git commit", Self::git_commit()),
            ("commit timestamp", env!("VERGEN_GIT_COMMIT_TIMESTAMP")),
            // build
            ("target triple", env!("VERGEN_CARGO_TARGET_TRIPLE")),
            // config
            ("Zcash network", (&config.network.network).into()),
```

**Review by teor2345 — APPROVED** [review]
Looks good, let's merge if CI passes.

**Review by teor2345 — CHANGES_REQUESTED** [review]
Looks like we might need a `metrics-exporter-prometheus` update before we can merge this change:
```
error: failed to select a version for `cc`.
    ... required by package `libgit2-sys v0.12.18+1.1.0`
    ... which is depended on by `git2 v0.13.17`
    ... which is depended on by `vergen v5.1.1`
    ... which is depended on by `zebrad v1.0.0-alpha.6 (/zebra/zebrad)`
versions that meet the requirements `^1.0.43` are: 1.0.67, 1.0.66, 1.0.65, 1.0.64, 1.0.63, 1.0.62, 1.0.61, 1.0.60, 1.0.59, 1.0.58, 1.0.57, 1.0.56, 1.0.55, 1.0.54, 1.0.53, 1.0.52, 1.0.51, 1.0.50, 1.0.49, 1.0.48, 1.0.47, 1.0.46, 1.0.45, 1.0.44, 1.0.43

all possible versions conflict with previously selected packages.

  previously selected package `cc v1.0.41`
    ... which is depended on by `generator v0.6.23`
    ... which is depended on by `loom v0.3.6`
    ... which is depended on by `sharded-slab v0.1.0`
    ... which is depended on by `metrics v0.13.0-alpha.8 (https://github.com/ZcashFoundation/metrics?rev=971133128e5aebe3ad177acffc6154449736cfa2#97113312)`
    ... which is depended on by `metrics-exporter-prometheus v0.1.0-alpha.7 (https://github.com/ZcashFoundation/metrics?rev=971133128e5aebe3ad177acffc6154449736cfa2#97113312)`
    ... which is depended on by `zebrad v1.0.0-alpha.6 (/zebra/zebrad)`
```

File: zebrad/src/application.rs (hunk 2 above)
Comment (fanatid) [reply zebrad/src/application.rs]: Thanks 👍

**Comment by teor2345** [comment]
> Looks like we might need a `metrics-exporter-prometheus` update before we can merge this change:

Here's what I tried:

Merge the most recent version of metrics-exporter-prometheus that *doesn't* use tokio-1.0 with the tokio-0.3 changes in the Foundation's metrics fork:
https://github.com/teor2345/metrics/tree/tokio-0.3%2Bmetrics-exporter-prometheus-v0.1.0-alpha.12

Update Zebra to use that branch:
https://github.com/teor2345/zebra/tree/vergen

But I get a metrics exporter acceptance test failure - it doesn't provide any output.

We had this issue once before when there was a version mismatch in some dependencies, so it should be fixable. But I might not have time to look at it for a few days.

**Comment by teor2345** [comment]
> We had this issue once before when there was a version mismatch in some dependencies, so it should be fixable. But I might not have time to look at it for a few days.

It was #1349, duplicate metrics dependencies. Seems like my attempt had similar issues.

**Comment by fanatid** [comment]
We do not need to update `metrics-exporter-prometheus` for the new `vergen`.

**Review by teor2345 — APPROVED** [review]
Looks like merging the latest main branch fixed it all up.

**Comment by dconnolly** [comment]
Looks like this is broken inside a `gcloud build` because the source is copied into a container image build and the git info injected as env vars, with no .git or associated env data present: https://github.com/ZcashFoundation/zebra/pull/2037

**Pull request merged**